ArjunaJTA/jta/classes/com/arjuna/ats/internal/jta/transaction/arjunacore/jca/XATerminatorImple.java

```diff
@@ -612,6 +612,9 @@ public Transaction getTransactionById(Object id) {
     }
 
     public Object getCurrentTransactionId() {
+        if (TransactionImple.getTransaction() == null)
```

Comment:
I know I did not mention it in the initial review but TransactionImple.getTransaction() can do a lot of work so I think you need to cache the result in a local variable in order to avoid the second call to getTransaction.